site/content/blog/identify-ghost-diff-during-kc-reconcile.md

```diff
@@ -0,0 +1,276 @@
+
+kapp, another carvel tool, tries to do a diff of the current live state of the resources with the desired state. Sometimes, even though there is no change in the desired state by the user explicitly, kapp still thinks some resources have changed and tries to redeploy them as part of the reconciliation. We call them `ghost` diffs. `Ghost` diffs can appear due to different reasons: controller can change some configuration of the resource, `HorizontalPodAutoscaler` influencing the no. of replicas for a deployment, etc. Everytime a `ghost` diff is detected, a new configMap gets created. kapp store its exit status and summary of operations(e.g. no. of updated/deleted/created resources) done in this configmap. If you have many packages creating `ghost` diffs in the Kubernetes(K8s) cluster, you will end up having large no. of unnecessary configMaps which can slow the K8s API server response time.
+
+Since Package consumers are aware of the `Package` configuration only, it becomes difficult for them to identify which part of the underlying resource configuration is causing these diffs.
+
+In this blog, we will see how to identify the resources causing these ghost diffs and also what part of their configuration is participating in it.
+
+
+## Prerequisites
+
+* carvel tool set
+* Kubernetes cluster(I'm using minikube)
+
+Ensure that kapp-controller is installed on your K8s cluster. Since this blog will be using `HorizontalPodAutoscaler`, we have to enable `metrics-server` on minikube.
+
+```bash
+$ minikube addons enable metrics-server
+    ▪ Using image k8s.gcr.io/metrics-server/metrics-server:v0.4.2
+  The 'metrics-server' addon is enabled
+```
+
+I will be using [kctrl](https://carvel.dev/kapp-controller/docs/latest/install/#installing-kapp-controller-cli-kctrl) to interact with `kapp-controller` resources. 
+
+## Install the Package
+
+For the purpose of this blog, I have already created a Carvel package `simple-app-package`. This Package is part of the package repository `my-pkg-repo`.  If you are interested in how to create package and package repository, I would recommend to go over [packaging-tutorial](https://carvel.dev/kapp-controller/docs/latest/packaging-tutorial/).
+
+First, we need to install the Package Repository. 
+
+```bash
+$ kctrl package repository add -r demo-pkg-repo --url docker.io/rohitagg2020/my-pkg-repo:1.0.0
+Target cluster 'https://192.168.64.82:8443' (nodes: minikube)
+
+Waiting for package repository to be added
+
+1:38:45PM: packagerepository/demo-pkg-repo (packaging.carvel.dev/v1alpha1) namespace: default: Reconciling
+1:38:50PM: packagerepository/demo-pkg-repo (packaging.carvel.dev/v1alpha1) namespace: default: ReconcileSucceeded
+
+Succeeded
+```
+
+Once the Package Repository is installed, we can check the list of available Packages.
```

Comment:
no need to be capitalizing words like "Package Repository" or "Package".

Reply:
Adopted